### ISSUE TITLE
Rule proposal: `component-name-in-template-casing`

### ISSUE THREAD

**Issue by michalsnik** [vuejs/eslint-plugin-vue#250]
**Style guide:**
https://vuejs.org/v2/style-guide/#Component-name-casing-in-templates-strongly-recommended

**Description:**
This rule would ensure proper component names in templates. As we only check `.vue` files in terms of templates we can provide both options: `kebab-case` and `PascalCase`

**Good examples:**
- kebab-case
```
<the-component />
```

- PascalCase
```
<TheComponent />
```

**Bad examples:**
```
<theComponent />
<Thecomponent />
<The-component />
```
and obviously with `kebab-case` option it would warn about `PascalCase` and vice-versa. 

Default option would be `PascalCase`.

**Comment by ota-meshi** [comment]
I'm sorry. I made a mistake in refs at PR 402 m(_ _)m

**Comment by Mouvedia** [comment]
> Default option would be PascalCase.

It shouldn't be the default.
cf https://stackoverflow.com/questions/22545621/do-custom-elements-require-a-dash-in-their-name

BTW it could be interesting to add a rule to warn about these:
- annotation-xml
- color-profile
- font-face
- font-face-src
- font-face-uri
- font-face-format
- font-face-name
- missing-glyph

**Comment by probil** [comment]
@Mouvedia, It should https://vuejs.org/v2/style-guide/#Component-name-casing-in-templates-strongly-recommended
And vue components are not custom elements

**Comment by Mouvedia** [comment]
@probil The rule is called `component-name-in-template-casing` and the style guide that you have linked just above support the kebab-case for DOM templates and "everywhere" else _but_ single file components.
It concludes with:

> **using kebab-case everywhere is also acceptable**

Why? Because using PascalCase in SFC and  kebab-case everywhere else would be inconsistent: we should recommend and support a solution that doesn't require to make that distinction.

**Comment by probil** [comment]
@Mouvedia Agree with you. `PascalCase` is not the best option for everyone. But the same for `kebab-case`. 
I don't think this eslint plugin is able to parse DOM templates at all. Because currently it relies on `*.vue` extension and `// @vue/component` comment so only SFC will be checked. But DOM templates are usually hidden somewhere in the back-end templates. _Correct me please, if I'm wrong here._

> In most projects, component names should always be PascalCase in single-file components

This plugin tries to follow official Vue Style Guide so `PascalCase` should be default. **But** of course. there should be configuration for this rule. So you can easily configure it to use `kebab-case` or any other approach you like.

**Comment by Mouvedia** [comment]
It depends on what matters the most to you:

- one is official and recent (september 2017)
- the other has precedence and wide spread (january 2017)

Both are valid options and have CONs/PROs.

**Comment by milewski** [comment]
This is not respecting the reserved vue `component` tag with the latest version of `"vue-eslint-parser": "^3.3.0"`

this: 
```html
<template>
    <component :is="element"></component>
</template>
```
is being transformed to this:
```html
<template>
    <Component :is="element"></Component>
</template>
```

even having the config to ignore `component` doesnt work:

```json
"vue/component-name-in-template-casing": [
      2, "PascalCase",
      {
        "ignores": ["component"]
      }
],
```
